aten/src/ATen/native/cuda/MultinomialKernel.cu

```diff
@@ -2,13 +2,506 @@
+      scalar_t curBucket = smem[threadIdx.x] + prevHighProb;
+      scalar_t prevBucket =
+          threadIdx.x == 0 ? prevHighProb :
+          smem[threadIdx.x - 1] + prevHighProb;
+      bool inBucket =
+          (cat < categories) &&
+          (!(sample >= curBucket) &&
+           (sample >= prevBucket) &&
+           (dist_val > zero));
+
+      if (inBucket) {
+        // We're done; we have the sample
+        // Torch indices are 1-based
+        dest[curDist] = cat;
+        found = true;
+      }
+
+      // Store the previous scan's high value for future use
+      prevHighProb = prevHighProb + smem[blockDim.x - 1];
+
+      __syncthreads();
+    }
+
+    if (threadIdx.x == 0 && !found) {
+      // This should address a rare bug where we don't select a valid index. This likely occurs when
+      // due to floating point arithmetic rounding errors, our cumulative sum does not add up to 1, but
+      // and our uniform sample is greater than this value. In this case we likely have unitialized memory
+      // in dest[curDist]. So basically we will loop through the distribution and pick the largest index
+      // where the distribution is non-zero. This is obviously terribly inefficient, but due to the
+      // rarity in which this occurs, this should not be an issue.
+      for (int cat = categories - 1; cat >= 0; --cat) {
+        if (dist[curDist * stride_dist + cat * stride_categories] > zero) {
+          dest[curDist] = cat;
+          break;
+        }
+      }
+    }
+  }
+}
+
+
+void multinomial_kernel_impl(Tensor& result, const Tensor& self, const int64_t n_sample, const bool with_replacement, Generator* generator) {
+  auto gen = get_generator_or_default<CUDAGenerator>(generator, cuda::detail::getDefaultCUDAGenerator());
+
+  int inputSize = self.dim();
+  int64_t numDist =
+      inputSize == 1 ? 1 : self.size(0);
+  int numCategories =
+      inputSize == 1 ? self.size(0) : self.size(1);
+
+  // Restructure data for 2d
+  if (inputSize == 1) {
+    self.unsqueeze_(0);
+  }
+
+  result.resize_({numDist, n_sample});
+
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "multinomial_kernel_cuda", [&] {
+    using accscalar_t = at::acc_type<scalar_t, true>;
+    auto props = at::cuda::getCurrentDeviceProperties();
+    assert(props != NULL);
+    int numSM = props->multiProcessorCount;
+    int maxThreads = props->maxThreadsPerBlock;
+    int maxShared = props->sharedMemPerBlock;
+    int requiredShared = (numCategories < maxThreads ? numCategories : maxThreads)
+                         * (sizeof(scalar_t) + sizeof(accscalar_t));
+
+    if (n_sample == 1 && maxShared >= requiredShared) {
+      // Optimized allocation-free implementation
+      // To exploit greater parallelism for the sampling, generate the
+      // Uniform random samples in a separate kernel launch, into
+      // temporarily allocated memory. The device RNG is thread-limited
+      Tensor sampled = at::empty({numDist, n_sample}, self.options());
```

Comment:
Better to call `native::empty_cuda` to avoid redispatch

Reply:
Fixed